Refactor students page fetching data through Gutenberg and avoiding subqueries

renatho opened the pull request:
### Changes proposed in this Pull Request

* It removes the subquery from the students query in order to make it work in environments with optimizations where the tables can live in different places.
* It removes an enqueued script that wasn't being used.
* It refactors the courses fetch in order to optimize and simplify using the `useSelect`.
* It introduces a new hook `useSelectWithDebounce` in order to expand the `useSelect`, to work with debounce.
* It removes a middleware we were using to change our rest API arguments to query strings. It seems it's working properly without the middleware with the URL rewrite enabled or not.
* It tweaks the `wp_kses` we use for the Students table, in order to support it in different envs that have customization on the `wp_kses`, like removing the `data-*`.

### Testing instructions

<!--
If the functionality of this PR is behind a feature flag, please include the
following testing step, using the correct name for the feature flag. (If you
aren't sure, just ignore this step)

* Enable the feature flag: `add_filter( 'sensei_feature_flag_{THE_FLAG_NAME}', '__return_true' );`
-->
* Create some courses and some students.
* Make some progress with some users.
* Visit WP admin > Sensei LMS > Students, and make sure it's listed properly as previously, especially the _"Last Activity"_ column.
* Add and remove a student from a course, and make sure it works as previously.

renatho committed:
Separate learners subquery

renatho committed:
Fix PHPCS

m1r0 reviewed (dismissed):
Looks good to me and works well. 👍

renatho committed:
Remove unused enqueue

renatho commented:
@m1r0, thank you for your review! But I realized that there are some other issues. I'll try to fix all related issues in this same PR.
I also added a new commit now to remove an enqueue script that we weren't using: https://github.com/Automattic/sensei/pull/5104/commits/3397c4ac519ae0e38e50083c35bd2b02a4d45848

renatho committed:
Add hook to handle useSelect with debounce

renatho committed:
Refactor course list to use core store

renatho committed:
Merge branch 'trunk' of github.com:Automattic/sensei into change/students-subquery

renatho commented:
Changing to draft again since I need to fix one other related thing.

gabrielcaires reviewed (commented):
Thanks @renatho for this refactoring,  I was already planning to refactor it to support new features.

Could you please explain how adding a `wordpress/data`  avoid subqueries?
One of our reasons to not use `wordpress/data` was because the sensei `wordpress/data` version is pretty outdated (2 years ago) so are avoiding to create new features that doesn't require a strong state management until we update it. 

Do you think it is really necessary to solve the problem?

I saw you are removing nock, usually mock the `fetch`  @wordpress/data library is not a good practice because it creates a coupling between the test and the implementation.
It already helped us to get some issues.

renatho committed:
Remove apiFetch middleware that changed the URL arguments to queryStrings

gabrielcaires reviewed (changes requested):
As we discussed, the main motivation of this PR on the frontends ide is to remove the http-client. 

I would suggest remove the changes related to ~@wordpress/data because we are using @wordpress/apiFetch that is also a standard way to load data from wordpress.
it avoid us to use a store and all related infrastructure to support `wordpress/data`.

What do you think?

renatho committed:
Add div custom kses for specific envs

renatho commented:
> Could you please explain how adding a wordpress/data avoid subqueries?

They are unrelated changes. I uptated the PR title in order to make it more clear. 😉 

> One of our reasons to not use wordpress/data was because the sensei wordpress/data version is pretty outdated (2 years ago) so are avoiding to create new features that doesn't require a strong state management until we update it.

Just registering here in the PR, as we talked, the version we have installed is for the unit tests only. When running the site, it will import it from from the Gutenberg bundled version in WordPress core: https://developer.wordpress.org/block-editor/contributors/versions-in-wordpress/. More details in: https://www.npmjs.com/package/@wordpress/dependency-extraction-webpack-plugin

> I would suggest remove the changes related to ~@wordpress/data because we are using @wordpress/apiFetch that is also a standard way to load data from wordpress.
it avoid us to use a store and all related infrastructure to support wordpress/data.

I was thinking a little about that. I think the only benefit of not using wordpress/data is not including their dependencies in the student page, but it's not a big deal since we're in wp-admin and it's used in many other pages, caching it separately. And using the useSelect there with the store, it brings many benefits, like using an implementation that is specifically for this purpose, saving some code writing and being more prepared for future changes.

So I'd say that it would be nice to continue with wordpress/data, unless you have a strong reason to not use it. WDYT?

gabrielcaires commented on assets/react-hooks/use-select-with-debounce.js:
```diff
@@ -0,0 +1,39 @@
+import { useSelect } from '@wordpress/data';
+import { useState, useEffect, useCallback } from '@wordpress/element';
+
+// TODO: Write tests.
```
Are you going to add this test?

renatho replied:
A partial test added here: https://github.com/Automattic/sensei/pull/5104/commits/859125c6185f321bb15a9fd343b20d090c0de92e, but I couldn't make the debounce work yet. So I'll keep this part as TODO for now. 😉

gabrielcaires commented on assets/admin/students/student-modal/index.test.js:
```diff
@@ -9,6 +9,12 @@ import {
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
```
Suggestion, replace `useSelect` to  `useEntityRecords` 
`const { records, isResolving } = useEntityRecords( 'postType', 'courses' );`

renatho replied:
Thank you for the suggestion! It would be nice, but this hook was exposed recently, so we can't use that yet in order to support older versions. See it in: https://github.com/WordPress/gutenberg/commit/8b8886c1dcd9de3a10d89a34c26af506ca5a8332

gabrielcaires reviewed (dismissed):
@renatho I still think that configuring a store just to render a list inside a modal (it is not a SPA)  just to make a get a bit overkill but considering it was moved to a hook if we can easy switch if we saw some issue with this approach.

I just noticed your add a TODO to write the hook test. Are you planning add this test?

renatho committed:
Add some tests to useSelectWithDebounce

The test is not complete, since we're having an issue to simulate the debounce timers.

aaronfc commented on assets/react-hooks/use-select-with-debounce.test.js:
```diff
@@ -0,0 +1,44 @@
+		rerender( <TestComponent deps={ deps2 } /> );
+		expect( mockFn ).toBeCalledWith( mapSelect, deps1 );
+
+		// TODO: Complete test running debounce to make sure it was called after the time.
+		// jest.runAllTimers();
+		// expect( mockFn ).toBeCalledWith( mapSelect, deps2 );
```
Do you plan to do this before merging?

renatho replied:
Unfortunately, I don't plan to add it now. We're having some problems with dependencies in order to make it work. `runAllTimers` is starting an infinite loop in the `debounce`. There are some issues about that on the Internet and it was supposed to be fixed, so at some point, we need to dig into the dependencies to see what's happening.

renatho replied:
Just adding some links here for future reference about this:

https://github.com/facebook/jest/issues/3465
https://github.com/lodash/lodash/issues/2893
https://github.com/facebook/jest/pull/7776

aaronfc commented on includes/class-sensei-db-query-learners.php:
```diff
@@ -105,6 +99,41 @@ private function build_query() {
+
+		$in_placeholders = implode( ', ', array_fill( 0, count( $user_ids ), '%s' ) );
+
+		// phpcs:disable WordPress.DB.DirectDatabaseQuery.NoCaching, WordPress.DB.DirectDatabaseQuery.DirectQuery
+		$results = $wpdb->get_results(
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare -- Placeholders created dinamically.
```
Just writing to confirm that you are using `phpcs:disable` on purpose. As opposed to `phpcs:ignore` this will disable the check until the end of the current file not just on the following line.

renatho replied:
Good catch! It wasn't on purpose. Fixed in: https://github.com/Automattic/sensei/pull/5104/commits/4e602d0ff36fc2eb24a110852c0b4b3b7bbcafc5

aaronfc reviewed (dismissed):
Looks good and works well 👍

renatho committed:
Fix PHPCS ignores

renatho commented:
After the last approval, I just added a commit fixing PHPCS comments: https://github.com/Automattic/sensei/pull/5104/commits/4e602d0ff36fc2eb24a110852c0b4b3b7bbcafc5. We need this PR in the next release, so considering this change is simple and doesn't impact how the plugin work, I'm going ahead and merging this. Please, let me know if there's something else on that, so I can raise a new PR. 😉
